### PR TITLE
bug: improve app exception handling

### PR DESCRIPTION
**Pull request by bradmartin**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Hardware (schematic, board, system design) change
- [ ] Software change

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have added / updated the documentation related to this change via either README or WIKI

### Software
- [ ] I have added tests to cover my changes.
- [ ] I have updated the `.github/workflows/build.yml` file to add my new test to the automated cloud build github action.
- [ ] All new and existing tests passed.
- [ ] My code follows the code style of this project.

### Hardware
- [ ] I have updated the design files (schematic, board, libraries).
- [ ] I have attached the PDFs of the SCH / BRD to this PR
- [ ] I have updated the design output (GERBER, BOM) files.

### CODE REVIEW

**Review by gcastro103 — APPROVED** [review]

File: apps/wear/smartdrive/app/pages/main-page/main-view-model.ts [hunk 1]
```diff
@@ -1235,15 +1240,19 @@ export class MainViewModel extends Observable {
     Application.on(
       Application.uncaughtErrorEvent,
       (args: UnhandledErrorEventData) => {
```
Comment (bradmartin) [inline apps/wear/smartdrive/app/pages/main-page/main-view-model.ts:1242]: This is one of the events where the app possibly crashes. Looks like we didn't have the `fullStop` being called. I did force the app to crash after a 10 second delay, without this code and the motor kept going. I added this method call in the exception handler and the motor of the SD did stop how it should.

This should improve the issue we get sometimes with users where the app crashes and the motor keeps going. If we don't handle the error the app should crash, they're very rare so we don't always get reports of it and it's been hard to reproduce becuase all we were told was: the app crashed and we also hadn't looked into the code in a year + to review it.

As i was fixing a bug for DCR-2207 to fix the watchface of the SmartDrive MX2+ app I looked at this issue and found that is likely the best course of action to fullStop on uncaught exceptions which would be the app crashing on android.

Any questions let me know.

Comment (guo-max) [reply apps/wear/smartdrive/app/pages/main-page/main-view-model.ts:1242]: I think the logic here make sense. But I do not know the details about the `fullStop` function. What if the `crashes` cause the `fullStop` not able to successfully execute? 
One suggestion is `kill` the `BLE` connection here as well. When `BLE` disconnected, the SD will stop the motor.

Comment (bradmartin) [reply apps/wear/smartdrive/app/pages/main-page/main-view-model.ts:1242]: fullstop kills everything that we can, it should always execute. This event ties into the android uncaught exception event so the code should execute.

``` typescript
 private _fullStop(forceDisable: boolean = false) {
    if (!this._ensuringBluetoothCapabilities || forceDisable) {
      // fullStop is called by app event handlers (such as suspend)
      // which may be triggered when we ask for permissions or to
      // enable the BLE radio when turning power assist on, so we
      // don't *always* want to turn power assist off
      if (this.powerAssistActive) {
        this.disablePowerAssist();
      }
    }
    if (this.isTraining) {
      this.onExitTrainingModeTap();
    }
  }
```

`disablePowerAssist()` is the guts

```typescript
async disablePowerAssist() {
    // ensure that we only vibrate if this was called while power
    // assist was actually on or the motor was still registered as
    // being on.
    const shouldVibrate = this.powerAssistActive || this.motorOn;

    // update state variables
    this.powerAssistActive = false;
    this.motorOn = false;

    sentryBreadCrumb('Disabling power assist');

    // make sure to stop any pending taps
    this._stopTaps();

    // clear timeout if there was one
    this._clearPowerAssistTimeout();

    // decrease energy consumption
    this._disableTapSensor();
    this._releaseCPU();
    this._enableWifi();
    this.powerAssistState = PowerAssist.State.Inactive;

    if (shouldVibrate) {
      // vibrate twice
      this._vibrator.vibrate([0, 200, 50, 200]);
    }

    // turn off the smartdrive
    try {
      await this._disconnectFromSmartDrive();
    } catch (err) {
      Sentry.captureException(err);
    }

    // reset our work interval
    if (this.chargingWorkTimeoutId === null) {
      this.chargingWorkTimeoutId = setInterval(
        this._doWhileCharged.bind(this),
        this.CHARGING_WORK_PERIOD_MS
      );
    }

    // now that we've disabled power assist - make sure the charts
    // update with the latest data from the smartdrive
    this._updateChartData();

    // now that we've updated the chart data and such - check the
    // activity for notifications
    this._updateNotifications();

    //now update the battery data information to send to the server

    this.BATTERY_DATA.stop(
      this.watchCurrentBatteryPercentage * 0.01,
      this.smartDriveCurrentBatteryPercentage,
      this.smartDrive.coastDistance,
      this.smartDrive.driveDistance
    );

    // no need to await this just watch for rejection
    this._kinveyService
      .sendBatteryData(this.BATTERY_DATA.toObj())
      .catch(err => {
        Sentry.captureException(err);
      });

    return Promise.resolve();
  }
```

In my 3 test runs it worked every time I forced the app to crash after 10 seconds and I had the motor running.

**Review by p-ranav — COMMENTED** [review]
Approved.

**Pull request merged**